### PR TITLE
ggml-cuda : use graph allocator

### PR DESCRIPTION
**Pull request by slaren**
Instead of allocating the tensors by ggml-cuda, the VRAM address assigned to the offloaded tensors is calculated from the offset that the graph allocator assigned to the tensor. It should result in a reduction in the size of the VRAM scratch buffer in all cases.

Removes scratch buffers from llama.cpp. This should make supporting other models easier. It should also allow removing the 512 batch size limit in `common.cpp`.

This is a quick hack. The proper solution will come with the unified backends interface, which will overhaul the memory management of the backends.

### CODE REVIEW

**Comment by Dampfinchen** [comment]
I can indeed confirm, this allows me to set 2 more layers to the GPU with the same VRAM usage. The speed increase is very welcome!

before
after

Awesome work. 13B on my RTX 2060 laptop gets faster and faster! It's incredible to me that we've come so far.

File: ggml-cuda.cu [hunk 1]
```diff
@@ -6308,16 +6313,48 @@ void ggml_cuda_assign_buffers_impl(struct ggml_tensor * tensor, bool scratch, bo
...
+
+    struct ggml_tensor_extra_gpu * extra = ggml_cuda_alloc_temp_tensor_extra();
+
+    const bool inplace = (tensor->src[0] != nullptr && tensor->src[0]->data == tensor->data) ||
+        tensor->op == GGML_OP_VIEW;
+
+    if (inplace && (tensor->src[0]->backend == GGML_BACKEND_GPU || tensor->src[0]->backend == GGML_BACKEND_GPU_SPLIT)) {
+        struct ggml_tensor_extra_gpu * src0_extra = (ggml_tensor_extra_gpu * ) tensor->src[0]->extra;
+        char * src0_ddc = (char *) src0_extra->data_device[g_main_device];
+        size_t offset = 0;
```
Comment (JohannesGaessler) [inline ggml-cuda.cu:6332]: This variable is shadowing the function argument. I think it should be renamed.

**Review by JohannesGaessler — COMMENTED** [review]
I can confirm that the PR reduces VRAM usage. I used the `perplexity` binary with a context size of 4096 for testing and got the VRAM usage from `nvidia-smi`:

| Model   | Test | Batch size | VRAM master [MiB] | VRAM PR [MiB] | VRAM saved [MiB] | t/s PR |
|---------|------|------------|-------------------|---------------|------------------|--------|
| 7b q4_0 | pp   |        256 | 6942              |          6836 | 106              |   1337 |
| 7b q4_0 | pp   |        512 | 6856              |          6634 | 222              |   1460 |
| 7b q4_0 | pp   |       1024 | -                 |          6978 | -                |   1611 |

Increasing the batch size beyond 512 seems to still make prompt processing faster at the cost of higher VRAM usage. However, when I set the batch size to 2048 I got the following error: `CUDA error 9 at ggml-cuda.cu:5823: invalid configuration argument`

**Comment by JohannesGaessler** [comment]
The issue at large batch sizes seems to be caused by RoPE:

```
(gdb) p *src0
$2 = {type = GGML_TYPE_F32, backend = GGML_BACKEND_GPU, n_dims = 3, ne = {128, 32, 2048, 1}, nb = {4, 512, 16384, 33554432}, op = GGML_OP_RESHAPE, 
  op_params = {0, 0, 0, 0, 0, 0, 0, 0}, is_param = false, grad = 0x0, src = {0x7fffa2e3e2e0, 0x0, 0x0, 0x0, 0x0, 0x0}, perf_runs = 1, 
  perf_cycles = 0, perf_time_us = 0, data = 0x7ffc34002020, name = "tmpk (reshaped)", '\000' <repeats 32 times>, extra = 0x5555572541a0, 
  padding = "\000\000\000"}
(gdb) p *src1
Cannot access memory at address 0x0
(gdb) p *dst
$3 = {type = GGML_TYPE_F32, backend = GGML_BACKEND_GPU, n_dims = 3, ne = {128, 32, 2048, 1}, nb = {4, 512, 16384, 33554432}, op = GGML_OP_ROPE, 
  op_params = {0, 128, 0, 0, 1176256512, 1065353216, 0, 0}, is_param = false, grad = 0x0, src = {0x7fffa2e3e540, 0x0, 0x0, 0x0, 0x0, 0x0}, 
  perf_runs = 0, perf_cycles = 0, perf_time_us = 0, data = 0x7ffc34002020, name = "Kcur", '\000' <repeats 43 times>, extra = 0x5555572542a0, 
  padding = "\000\000\000"}
```

File: llama.cpp [hunk 2]
```diff
@@ -1293,15 +1196,15 @@ static void llama_model_load_internal(
         (void) n_batch;
 #ifdef GGML_USE_CUBLAS
         if (low_vram) {
-            LLAMA_LOG_INFO("%s: not allocating a VRAM scratch buffer due to low VRAM option\n", __func__);
+            LLAMA_LOG_INFO("%s: (debug) not allocating a VRAM scratch buffer due to low VRAM option\n", __func__);
```
[the remark this answers is not in the document]
Comment (slaren) [reply llama.cpp:1199]: This is just for comparing the memory usage with master, it will be removed.

**Comment by slaren** [comment]
I can use batch sizes above 2048, but it's not any faster (3090Ti wsl2):
| model                          | backend    | n_gpu_layers |  n_threads |    n_batch | test       |             t/s |
| ------------------------------ | ---------- | -----------: | ---------: | ---------: | ---------- | --------------: |
| LLaMA 7B mostly Q4_0           | CUDA       |           99 |         16 |        512 | pp 512     | 2233.33 ± 23.64 |
| LLaMA 7B mostly Q4_0           | CUDA       |           99 |         16 |       1024 | pp 512     |  2245.00 ± 5.12 |
| LLaMA 7B mostly Q4_0           | CUDA       |           99 |         16 |       2048 | pp 512     |  2243.52 ± 7.39 |
| LLaMA 7B mostly Q4_0           | CUDA       |           99 |         16 |       4096 | pp 512     |  2245.93 ± 8.25 |

build: 188a285 (1011)

**Comment by slaren** [comment]
It looks like `block_nums` is `1, 131072, 1`, but the maximum per y or z dimension is 65535. However the x dimension should not have that limit.

**Comment by JohannesGaessler** [comment]
If there indeed is a maximum grid size in y and z direction only that would be very stupid because the CUDA documentation assured me that x, y, and z only make a difference conceptually for the programmer.

**Comment by slaren** [comment]
It seems to be the case. It was probably a decision made long ago to save 32 bits in the block indexes.

image

**Comment by JohannesGaessler** [comment]
I've tried a quick patch where I just swapped the x and y dimensions but this did **not** work to fix the maximum batch size.

**Comment by JohannesGaessler** [comment]
Disregard my previous post I forgot that there are multiple kernels that may need to be patched. `diag_mask_inf` seems to also be causing problems:

```
(gdb) p *src0
$1 = {type = GGML_TYPE_F32, backend = GGML_BACKEND_GPU, n_dims = 3, ne = {2048, 2048, 32, 1}, nb = {4, 8192, 16777216, 536870912}, 
  op = GGML_OP_SCALE, op_params = {0, 0, 0, 0, 0, 0, 0, 0}, is_param = false, grad = 0x0, src = {0x7fffa2e3f4b0, 0x7fffa2e3df50, 0x0, 0x0, 0x0, 
    0x0}, perf_runs = 1, perf_cycles = 0, perf_time_us = 0, data = 0x7ffc36002020, name = "KQ_scaled", '\000' <repeats 38 times>, 
  extra = 0x555557254fa0, padding = "\000\000\000"}
(gdb) p *src1
Cannot access memory at address 0x0
(gdb) p *dst
$2 = {type = GGML_TYPE_F32, backend = GGML_BACKEND_GPU, n_dims = 3, ne = {2048, 2048, 32, 1}, nb = {4, 8192, 16777216, 536870912}, 
  op = GGML_OP_DIAG_MASK_INF, op_params = {0, 1, 0, 0, 0, 0, 0, 0}, is_param = false, grad = 0x0, src = {0x7fffa2e3f5e0, 0x0, 0x0, 0x0, 0x0, 0x0}, 
  perf_runs = 0, perf_cycles = 0, perf_time_us = 0, data = 0x7ffc36002020, name = "KQ_masked", '\000' <repeats 38 times>, extra = 0x5555572550a0, 
  padding = "\000\000\000"}
```

**Comment by JohannesGaessler** [comment]
I've pushed a quick fix here: https://github.com/JohannesGaessler/llama.cpp/tree/cuda-graph-allocr
This only modifies the kernels that caused the issue, a proper fix would probably be to more generally use the x dimension in the direction in which tensors are being flattened though.

**Comment by slaren** [comment]
Can you push the fix to this branch too?

**Comment by JohannesGaessler** [comment]
This is the table with the additional results for 2048 and 4096 batch size:

| Model   | Test | Batch size | VRAM master [MiB] | VRAM PR [MiB] | VRAM saved [MiB] | t/s PR |
|---------|------|------------|-------------------|---------------|------------------|--------|
| 7b q4_0 | pp   |        256 | 6942              |          6836 | 106              |   1337 |
| 7b q4_0 | pp   |        512 | 6856              |          6634 | 222              |   1460 |
| 7b q4_0 | pp   |       1024 | -                 |          6978 | -                |   1611 |
| 7b q4_0 | pp   |       2048 | -                 |          7674 | -                |   1609 |
| 7b q4_0 | pp   |       4096 | -                 |          9158 | -                |   1408 |

I also don't observe any benefit from using a batch size > 1024.

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
Here is one more data point

ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 4080, compute capability 8.9
| model                          | backend    | n_gpu_layers |  n_threads |    n_batch | test       |             t/s |
| ------------------------------ | ---------- | -----------: | ---------: | ---------: | ---------- | --------------: |
| LLaMA 7B mostly Q4_0           | CUDA       |          100 |          1 |        256 | pp 2047    | 2584.35 ± 10.21 |
| LLaMA 7B mostly Q4_0           | CUDA       |          100 |          1 |        512 | pp 2047    |  2785.69 ± 8.58 |
| LLaMA 7B mostly Q4_0           | CUDA       |          100 |          1 |       1024 | pp 2047    |  2839.73 ± 8.40 |
| LLaMA 7B mostly Q4_0           | CUDA       |          100 |          1 |       2048 | pp 2047    |  2651.06 ± 0.87 |
| LLaMA 7B mostly Q4_0           | CUDA       |          100 |          1 |        256 | tg 128     |   130.11 ± 0.01 |
| LLaMA 7B mostly Q4_0           | CUDA       |          100 |          1 |        512 | tg 128     |   130.11 ± 0.06 |
| LLaMA 7B mostly Q4_0           | CUDA       |          100 |          1 |       1024 | tg 128     |   130.14 ± 0.02 |
| LLaMA 7B mostly Q4_0           | CUDA       |          100 |          1 |       2048 | tg 128     |   130.13 ± 0.01 |

build: 76515b7 (1015)

**Comment by staviq** [comment]
Hey, question: How does this relate to mul_mat_q ?

Because I'm not sure why, but I'm getting less VRAM usage and more performance, with -mmq -b 512, than without -mmq and with -b 128

( I have 2 asymmetric GPUs so the VRAM usage is guesstimated from the first GPU from nvidia-smi )

`-b 512 -mmq` ~ 7.8G VRAM

```
llama_print_timings:        load time =  7651.36 ms
llama_print_timings:      sample time =    11.66 ms /    23 runs   (    0.51 ms per token,  1972.39 tokens per second)
llama_print_timings: prompt eval time = 33378.26 ms /  1712 tokens (   19.50 ms per token,    **51.29** tokens per second)
llama_print_timings:        eval time =  2147.82 ms /    22 runs   (   97.63 ms per token,    10.24 tokens per second)
llama_print_timings:       total time = 35542.36 ms
```

`-b 128` ~ 7.9G VRAM

```
llama_print_timings:        load time =  7638.86 ms
llama_print_timings:      sample time =    14.90 ms /    29 runs   (    0.51 ms per token,  1946.18 tokens per second)
llama_print_timings: prompt eval time = 46737.82 ms /  1712 tokens (   27.30 ms per token,    **36.63** tokens per second)
llama_print_timings:        eval time =  2709.72 ms /    28 runs   (   96.78 ms per token,    10.33 tokens per second)
llama_print_timings:       total time = 49467.86 ms
```

I checked, and I'm on `commit 519c981f8b65ee6c87c2965539685ced0a17223b (HEAD -> master, origin/master, origin/HEAD)`

**Comment by slaren** [comment]
All this does is reduce the memory usage of the CUDA backend, it shouldn't have an effect on performance, except when using a larger batch size than 512, which previously was not possible.

**Pull request merged**